### PR TITLE
Channel group cleaner

### PR DESCRIPTION
**Pull request by mehmetalisavas**

### CODE REVIEW

File: go/src/socialapi/models/channel.go [hunk 1]
```diff
@@ -1162,3 +1165,64 @@ func (c *Channel) ShowUnreadCount() bool {
...
+				continue
+			}
+			// if error is not nil and equal to record not found
+			// then remove the channel and its participants in postgre
+			if err != nil && err == mgo.ErrNotFound {
+				if err = channel.DeleteWithParticipantsForce(); err != nil {
+					errs = multierror.Append(errs, err)
+				}
+			} else {
+				errs = multierror.Append(errs, err)
```
Comment (cihangir) [inline go/src/socialapi/models/channel.go]: you are going to append even if error is nil

File: go/src/socialapi/models/channel.go [hunk 2]
```diff
@@ -1162,3 +1165,64 @@ func (c *Channel) ShowUnreadCount() bool {
...
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+
+		for _, channel := range channels {
+			_, err := modelhelper.GetGroup(channel.GroupName)
+			// if group already exists in mongo, then we don't need to fecth the same data
+			// while fetching channels
+			if err == nil {
+				offset++
```
Comment (cihangir) [inline go/src/socialapi/models/channel.go]: offset needs to be increased even if there is an error

File: go/src/socialapi/models/channel.go [hunk 3]
```diff
@@ -1162,3 +1165,64 @@ func (c *Channel) ShowUnreadCount() bool {
...
+				}
+			} else {
+				errs = multierror.Append(errs, err)
+			}
+		}
+
+		// This check provide us to break the loop if there is no data left that need
+		// to be processed fetch tolerance is limit/2, if fetched channels count is
+		// less than limited number then break the loop.
+		if len(channels) < (limit / 2) {
```
Comment (cihangir) [inline go/src/socialapi/models/channel.go]: i think "limit / 2" is unnecessary, if it is smaller than len(channels) we can assume safe to break out of the loop

File: go/src/socialapi/models/channel.go (hunk 1 above)
Comment (mehmetalisavas) [inline go/src/socialapi/models/channel.go]: actually, we'r appending only the error except `mgo.ErrNotFound` errors. there is early return on above line while checking `if err==nil{continue}`

File: go/src/socialapi/models/channel.go (hunk 2 above)
Comment (mehmetalisavas) [inline go/src/socialapi/models/channel.go]: i thought it would be nice to try to fetch Group again. is it guarenteed to get the same error ?

Comment (cihangir) [inline go/src/socialapi/models/channel.go]: most probably yes

File: go/src/socialapi/models/channel.go (hunk 1 above)
Comment (cihangir) [inline go/src/socialapi/models/channel.go]: 🤦‍♂️ 👍

**Comment by mehmetalisavas** [comment]
@cihangir ping

File: go/src/socialapi/scripts/channelcleaner/channelcleaner_test.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	// In this case we need to try to create same named group in mongo.
+	// then it gives key duplication error, so err checking is disabled here.
+	models.CreateGroupInMongo(groupName, groupChannel.Id)
+	// So(err, ShouldBeNil)
+
+	return groupChannel, account
+}
+
+//
+// func accountCreatorWithCount(count int) error {
```
Comment (cihangir) [inline go/src/socialapi/scripts/channelcleaner/channelcleaner_test.go]: why this commented out code still here?

**Review by cihangir — CHANGES_REQUESTED** [review]
commented out code.

**Comment by cihangir** [comment]
please rebase you branch

**Review by cihangir — APPROVED** [review]

**Pull request merged**